### PR TITLE
Add first draft of top of tree dev proposal

### PR DESCRIPTION
**Pull request by neonichu**

### CODE REVIEW

File: proposals/xxxx-package-manager-top-of-tree.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+### Enhancements to `swift package edit`
+
+We will extend the `edit` subcommand with a new optional argument `--path`:
+
+```bash
+$ swift package edit bar --path ../bar
+```
+
+This will make `./Packages/bar` a symbolic link to the given path in the local filesystem and will ensure that the package manager will no longer be responsible for managing checkouts for the `bar` package, instead the user is responsible for managing the source control operations on their own. This is consistent with the current behaviour of `swift edit`. Using `swift package unedit` will also work unchanged.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-top-of-tree.md]: I think the symbolic link part is an implementation detail, we should first mention the problem this solves (and how). i.e. users usually have a checkout of the repository they work on and would like to attach that in this package graph.

We should also mention that the unedit part will not remove the contents at symlink and just remove the connection.

Also, if the path doesn't exist should we create a checkout?

File: proposals/xxxx-package-manager-top-of-tree.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+let package = Package(
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", branch: "development"),
+    ]
+)
+```
+
+Instead of checking out a tag, the package manager will check out the specified branch for the given package. If a package exists simultaneously with a specified version range and with a specified branch in the same package graph, an error will be emitted and checking out dependencies will fail. It is also an error to depend on a package referencing a branch transitively from a tagged package.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-top-of-tree.md]: I think we should allow if there are multiple references to a dependency and the root package contains the branch reference, and just emit warning in that case.

File: proposals/xxxx-package-manager-top-of-tree.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", branch: "development"),
+    ]
+)
+```
+
+Instead of checking out a tag, the package manager will check out the specified branch for the given package. If a package exists simultaneously with a specified version range and with a specified branch in the same package graph, an error will be emitted and checking out dependencies will fail. It is also an error to depend on a package referencing a branch transitively from a tagged package. 
+
+Running `swift package update` will update packages referencing branches to their latest remote state. Running `swift package pin` will store the commit hash for the currently checked out revision in the `Pinfile`, so that other users of the package will receive the exact same revision if pinning is enabled.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-top-of-tree.md]: We will also want to store the branch name in pins file

File: proposals/xxxx-package-manager-top-of-tree.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+- bootstrapping a new package which does not yet have a version at all
+- allowing developers to manually check out repositories on their machines as overrides -- this is useful when developing multiple packages in tandem or when working on packages alongside an application
+- allowing references to packages in branches instead of tags -- this is useful for working on revlocked packages
+
+## Proposed solution
+
+We propose two solutions for these problems:
+
+- extend the package manifest to allow specifying a branch instead of a version to support revlocked packages
+- extend `swift package edit` to take an optional path argument to manually override the current behaviour
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-top-of-tree.md]: This feels a little vague

File: proposals/xxxx-package-manager-top-of-tree.md (hunk 2 above)
Comment (neonichu) [reply proposals/xxxx-package-manager-top-of-tree.md]: Maybe we could do a sanity check against the git history? If the branch is ahead of the version tag other occurrences of the dependency reference, we warn, otherwise we emit an error?

File: proposals/xxxx-package-manager-top-of-tree.md (hunk 1 above)
Comment (neonichu) [reply proposals/xxxx-package-manager-top-of-tree.md]: Yah, I think creating the checkout on the user's behalf makes sense if it does not exist, yet.

**Comment by neonichu** [comment]
Split this into two proposals now as we originally intended.

File: proposals/xxxx-package-manager-branch-support.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
+# Package Manager Support for branches in the package manifest
+
+* Proposal: [SE-XXXX](xxxx-package-manager-branch-support.md)
+* Author: [Boris Bügling](https://github.com/neonichu)
+* Review Manager: TBD
+* Status: Discussion
+
+## Introduction
+
+This proposal adds enhancements to the package manifest to support development of packages without strict versioning. It is related to "top of tree" development, but described in a separate proposal, because the implementation is entirely orthogonal to it.
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: I would reword the second sentence to something like: "This is one of two features, along with <insert name of other proposal here>, being proposed to enable use of SwiftPM to develop on "top of tree" of related packages."

File: proposals/xxxx-package-manager-branch-support.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Package Manager Support for branches in the package manifest
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: I would leave off "in the package manifest", as that's a detail of our proposed solution, which is up for discussion. Maybe "Package Manager Support for Branch-Based Development"

File: proposals/xxxx-package-manager-branch-support.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+* Review Manager: TBD
+* Status: Discussion
+
+## Introduction
+
+This proposal adds enhancements to the package manifest to support development of packages without strict versioning. It is related to "top of tree" development, but described in a separate proposal, because the implementation is entirely orthogonal to it.
+
+## Motivation
+
+The package manager currently supports packages depedencies which are strictly versioned according to semantic versioning. This works well for users of packages, but we still see some use cases where this hinders common development workflows:
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: "Dependencies" is misspelled.
Instead of "This works well for users of packages...", I would say something like "This is how a package's dependencies should be specified when that package is released, but this requirement hinders some development workflows:

File: proposals/xxxx-package-manager-branch-support.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+- bootstrapping a new package which does not yet have a version at all
+- allowing references to packages in branches instead of tags -- this is useful for working on revlocked packages
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: Allowing references to branches is the solution, not the broken workflow. I would rephrase the 2nd workflow to something like "Developing related packages in tandem in between releases, when one package may depend on the latest revision of another, which has not yet been tagged for release."

File: proposals/xxxx-package-manager-branch-support.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+## Proposed solution
+
+As a solution to this problem, we propose to extend the package manifest to allow specifying a branch instead of a version to support revlocked packages and initial bootstrapping.
+
+## Detailed Design
+
+We will introduce a new initializer for `Package` which takes a branch instead of a version range:
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: Just to make it clear that this isn't replacing the prior initializer, I would call this a "second" initializer, not a "new" initializer. Also, this should be reworded slightly to make it clear that it's an initializer for the .Package dependency enum, not for the top level Package object.

File: proposals/xxxx-package-manager-branch-support.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+let package = Package(
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", branch: "development"),
+    ]
+)
+```
+
+Instead of checking out a tag, the package manager will check out the specified branch for the given package. If a package exists simultaneously with a specified version range and with a specified branch in the same package graph, an error will be emitted and checking out dependencies will fail. It is also an error to depend on a package referencing a branch transitively from a tagged package.
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: I would reword the first sentence to:
"Whenever dependencies are checked out or updated, if a dependency on a package specifies a branch instead of a version, the latest commit on that branch will be checked out for that package.

Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: I would reword the second sentence to:
"If a dependency on a package specifies a branch instead of a version range, all other dependencies on that package in the current package graph must specify the same branch, and not a version, or an error will be emitted and checking out dependencies will fail.

Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: The third sentence is hard to understand. I would rephrase, and move it into its own paragraph. Something like:
"While this feature is useful during development, a package's dependencies should be updated to point at versions instead of branches before that package is tagged for release. This is because a released package should provide a stable specification of its dependencies, and not break when a branch changes over time. To enforce this, it is an error if a package referenced by a version-based dependency specifies a branch in any of its dependencies. (Note that the converse is OK; a package referenced by a branch-based dependency may specify versions in its dependencies.)

File: proposals/xxxx-package-manager-branch-support.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", branch: "development"),
+    ]
+)
+```
+
+Instead of checking out a tag, the package manager will check out the specified branch for the given package. If a package exists simultaneously with a specified version range and with a specified branch in the same package graph, an error will be emitted and checking out dependencies will fail. It is also an error to depend on a package referencing a branch transitively from a tagged package. 
+
+Running `swift package update` will update packages referencing branches to their latest remote state. Running `swift package pin` will store the commit hash for the currently checked out revision in the `Pinfile`, as well as the branch name, so that other users of the package will receive the exact same revision if pinning is enabled.
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: Let's discuss if this is the behavior we want from pinning. Should autopinning affect branch-based dependencies?

File: proposals/xxxx-package-manager-branch-support.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+## Impact on existing code
+
+There will no impact on existing code.
+
+## Alternative considered
+
+None at this point.
```
Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: One alternative we should mention would be storing the branch in a sidecar file that overrides the manifest instead of storing it in the manifest itself.

Comment (rballard) [inline proposals/xxxx-package-manager-branch-support.md]: Another alternative would be to allow packages to be released while specifying branches for its dependencies. Though we did kind of address that above already (if you accept my proposed reword)

File: proposals/xxxx-package-manager-branch-support.md (hunk 11 above)
Comment (neonichu) [reply proposals/xxxx-package-manager-branch-support.md]: I think it definitely should, the risk of a branch-based dependency breaking when being updated is much higher. Also this would be consistent with how other package managers handle this, e.g. CocoaPods, Bundler and Carthage.

File: proposals/xxxx-package-manager-branch-support.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+let package = Package(
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", branch: "development"),
+    ]
+)
+```
+
+Whenever dependencies are checked out or updated, if a dependency on a package specifies a branch instead of a version, the latest commit on that branch will be checked out for that package. If a dependency on a package specifies a branch instead of a version range, all other dependencies on that package in the current package graph must specify the same branch, and not a version, or an error will be emitted and checking out dependencies will fail.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-branch-support.md]: I think we should allow other dependencies to have versions and if root package mentions a branch it overrides the versioned dependencies.

File: proposals/xxxx-package-manager-branch-support.md [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+While this feature is useful during development, a package's dependencies should be updated to point at versions instead of branches before that package is tagged for release. This is because a released package should provide a stable specification of its dependencies, and not break when a branch changes over time. To enforce this, it is an error if a package referenced by a version-based dependency specifies a branch in any of its dependencies.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-branch-support.md:65]: > To enforce this, it is an error if a package referenced by a version-based dependency specifies a branch in any of its dependencies.

Doesn't this contradicts the above point? ("all other dependencies on that package in the current package graph must specify the same branch")

File: proposals/xxxx-package-manager-branch-support.md [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+Running `swift package update` will update packages referencing branches to their latest remote state. Running `swift package pin` will store the commit hash for the currently checked out revision in the `Pinfile`, as well as the branch name, so that other users of the package will receive the exact same revision if pinning is enabled.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-branch-support.md]: nit: "pins file"

File: proposals/xxxx-package-manager-top-of-tree.md [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+This allows users to manage their own checkout of the `bar` repository and will make the package manager use that instead of checking out a tagged version as it normally would. Concretely, this will make `./Packages/bar` a symbolic link to the given path in the local filesystem and will ensure that the package manager will no longer be responsible for managing checkouts for the `bar` package, instead the user is responsible for managing the source control operations on their own. This is consistent with the current behaviour of `swift edit`. Using `swift package unedit` will also work unchanged, but the checkout itself will not be deleted, only the symlink. If there is no existing checkout at the given filesystem location, the package manager will do an initial clone on the user's behalf.
+
+## Impact on existing code
+
+There will no impact on existing code.
+
+## Alternative considered
+
+None at this point.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-top-of-tree.md]: 👍

File: proposals/xxxx-package-manager-branch-support.md (hunk 14 above)
Comment (neonichu) [reply proposals/xxxx-package-manager-branch-support.md:65]: 🤔 True, I think it makes sense to go with your suggestion for the previous paragraph and then this contradiction is resolved.

File: proposals/xxxx-package-manager-branch-support.md (hunk 13 above)
Comment (neonichu) [reply proposals/xxxx-package-manager-branch-support.md]: I thought this might be confusing, but on the other hand you're right that it contradicts with the next paragraph and also with the intent of the proposal, because this would mean having to change a bunch of package manifests if a dependency is also used by leaf packages.

File: proposals/xxxx-package-manager-branch-support.md (hunk 14 above)
Comment (aciidgh) [reply proposals/xxxx-package-manager-branch-support.md:65]: 👍

File: proposals/xxxx-package-manager-branch-support.md (hunk 13 above)
Comment (aciidgh) [reply proposals/xxxx-package-manager-branch-support.md]: 👍

File: proposals/xxxx-package-manager-branch-support.md [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+## Motivation
+
+The package manager currently supports packages dependencies which are strictly versioned according to semantic versioning. This is how a package's dependencies should be specified when that package is released, but this requirement hinders some development workflows:
+
+- bootstrapping a new package which does not yet have a version at all
+- developing related packages in tandem in between releases, when one package may depend on the latest revision of another, which has not yet been tagged for release
+
+## Proposed solution
+
+As a solution to this problem, we propose to extend the package manifest to allow specifying a branch instead of a version to support revlocked packages and initial bootstrapping.
```
Comment (ddunbar) [inline proposals/xxxx-package-manager-branch-support.md]: Did we decide not to accept a revision here? Why not?

File: proposals/xxxx-package-manager-branch-support.md [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+let package = Package(
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", branch: "development"),
+    ]
+)
+```
+
+Whenever dependencies are checked out or updated, if a dependency on a package specifies a branch instead of a version, the latest commit on that branch will be checked out for that package. If a dependency on a package specifies a branch instead of a version range, it will override any versioned dependencies present in the current package graph that other packages might specify.
```
Comment (ddunbar) [inline proposals/xxxx-package-manager-branch-support.md:55]: I think we should call out just how dangerous that last sentence is (that it is clobbering requirements from other packages), and the use that to lead into next paragraph.

Comment (ddunbar) [reply proposals/xxxx-package-manager-branch-support.md:55]: Specifically, what we would like to address here is situations where a package has changed their development dependencies in such a way that, for example, the graph is unresolvable, but they don't get an error because something is in branch mode.

File: proposals/xxxx-package-manager-branch-support.md [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+While this feature is useful during development, a package's dependencies should be updated to point at versions instead of branches before that package is tagged for release. This is because a released package should provide a stable specification of its dependencies, and not break when a branch changes over time. To enforce this, it is an error if a package referenced by a version-based dependency specifies a branch in any of its dependencies.
+
+Running `swift package update` will update packages referencing branches to their latest remote state. Running `swift package pin` will store the commit hash for the currently checked out revision in the pins file, as well as the branch name, so that other users of the package will receive the exact same revision if pinning is enabled.
```
Comment (ddunbar) [inline proposals/xxxx-package-manager-branch-support.md]: What will we use the branch name for here? Diagnostics?

File: proposals/xxxx-package-manager-branch-support.md [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+## Impact on existing code
+
+There will no impact on existing code.
+
+## Alternative considered
+
+Storing the branch in a separate file that overrides the manifest, instead of the manifest itself.
```
Comment (ddunbar) [inline proposals/xxxx-package-manager-branch-support.md]: We should elaborate on this alternative and say why we didn't take it.

We should elaborate a tad on why we made it an error to publish a package with this (this was a subject of debate).

File: proposals/xxxx-package-manager-top-of-tree.md [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+## Detailed Design
+
+We will extend the `edit` subcommand with a new optional argument `--path`:
+
+```bash
+$ swift package edit bar --path ../bar
+```
+
+This allows users to manage their own checkout of the `bar` repository and will make the package manager use that instead of checking out a tagged version as it normally would. Concretely, this will make `./Packages/bar` a symbolic link to the given path in the local filesystem and will ensure that the package manager will no longer be responsible for managing checkouts for the `bar` package, instead the user is responsible for managing the source control operations on their own. This is consistent with the current behaviour of `swift edit`. Using `swift package unedit` will also work unchanged, but the checkout itself will not be deleted, only the symlink. If there is no existing checkout at the given filesystem location, the package manager will do an initial clone on the user's behalf.
```
Comment (ddunbar) [inline proposals/xxxx-package-manager-top-of-tree.md]: I think we should discuss why we put the symbolic link down.

File: proposals/xxxx-package-manager-branch-support.md (hunk 17 above)
Comment (neonichu) [reply proposals/xxxx-package-manager-branch-support.md]: We do now.

File: proposals/xxxx-package-manager-branch-support.md [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+We decided to make publishing a package with unversioned dependencies an error, because a released package should provide a stable specification of its dependencies. A dependency on a branch could break at any time when the branch is being changed over time.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-branch-support.md]: This is confusing, we don't have publishing feature

File: proposals/xxxx-package-manager-branch-support.md [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+
+Running `swift package update` will update packages referencing branches to their latest remote state. Running `swift package pin` will store the commit hash for the currently checked out revision in the pins file, as well as the branch name, so that other users of the package will receive the exact same revision if pinning is enabled. If a revision was specified, users will always receive that specific revision and `swift package update` becomes a no op.
+
+### Pinning to a branch or revision
+
+In addition to specifying a branch or revision in the manifest, we will also allow specifying it when pinning:
+
+```bash
+$ swift pin --branch <branchname>
+$ swift pin --revision <revision>
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-branch-support.md]: We should mention somewhere that this will only take the 40 len hash and not other treeish

File: proposals/xxxx-package-manager-branch-support.md [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+While this feature is useful during development, a package's dependencies should be updated to point at versions instead of branches before that package is tagged for release. This is because a released package should provide a stable specification of its dependencies, and not break when a branch changes over time. To enforce this, it is an error if a package referenced by a version-based dependency specifies a branch in any of its dependencies.
+
+Running `swift package update` will update packages referencing branches to their latest remote state. Running `swift package pin` will store the commit hash for the currently checked out revision in the pins file, as well as the branch name, so that other users of the package will receive the exact same revision if pinning is enabled. If a revision was specified, users will always receive that specific revision and `swift package update` becomes a no op.
+
+### Pinning to a branch or revision
+
+In addition to specifying a branch or revision in the manifest, we will also allow specifying it when pinning:
+
+```bash
+$ swift pin --branch <branchname>
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-branch-support.md]: This should be `swift package pin <package-name> --branch <branch-name>`

File: proposals/xxxx-package-manager-branch-support.md [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", revision: "0123456789012345678901234567890123456789"),
+    ]
+)
+```
+
+Whenever dependencies are checked out or updated, if a dependency on a package specifies a branch instead of a version, the latest commit on that branch will be checked out for that package. If a dependency on a package specifies a branch instead of a version range, it will override any versioned dependencies present in the current package graph that other packages might specify.
+
+While this feature is useful during development, a package's dependencies should be updated to point at versions instead of branches before that package is tagged for release. This is because a released package should provide a stable specification of its dependencies, and not break when a branch changes over time. To enforce this, it is an error if a package referenced by a version-based dependency specifies a branch in any of its dependencies.
```
Comment (aciidgh) [inline proposals/xxxx-package-manager-branch-support.md:65]: We should add some examples of graphs which are valid and not valid

**Comment by neonichu** [comment]
Superseded by apple/swift-evolution#594

**Comment by chris-hatton** [comment]
Just a note to say it's great to see this feature coming in: it is critical to our team's ability to adopt the Swift Package Manager as we develop several projects along with their shared modules in parallel and want at least some of those dependencies to always remain 'latest'.

It would be even more useful if this feature could be driven by the branch-name or tag of the project repository i.e. if compiling from a branch named 'development', use the latest version of a dependency, and if compiling from a release-tagged or 'release/' prefixed branch, then use a pinned version of a dependency.

Or, would there be another emergent way of supporting this workflow?

**Comment by ddunbar** [comment]
@chris-hatton we discussed that case, where you want the branch to be driven by something at the higher level (either the higher level package or the tool itself). I think this SE will at least let you easily implement it yourself w/ project specific scripts, and then we can discuss what more features it is worth supporting directly in the tool... how does that sound to you?
